src/Uno.Toolkit.Skia.WinUI/Controls/Shadows/ShadowContainer.Paint.cs

```diff
@@ -133,12 +123,12 @@ private void OnSurfacePainted(object? sender, SKPaintSurfaceEventArgs e)
 			shape.DrawInnerShadow(state, canvas, paint, shadow);
 		}
 
-		// If a property has changed dynamically during this paint method,
-		// then we don't want to cache the updated shadows
-		if (!_isShadowDirty)
-		{
-			Cache.AddOrUpdate(key, e.Surface.Snapshot());
-		}
+		// No property should have changed during painting.
+#if DEBUG
+		global::System.Diagnostics.Debug.Assert(!_isShadowDirty);
+#endif
+
+		Cache.AddOrUpdate(key, e.Surface.Snapshot());
```

Comment:
there is no guarantee that this cant happen.
when it does, the latest key and the outdated snapshot will be paired,
causing an invalid cache to be created

Reply:
Why? My understanding is that from the moment `_isShadowDirty` is cleared earlier in the function, the only logic run is some drawing logic which doesn't (and shouldn't) change the state.
As a general concept, painting should basically be a "pure" function of the current state (except we change the canvas state and restore in the end, but whatever). Not doing so is most often unnecessary and it also opens you up to infinite invalidation bugs where painting invalidates the state, which leads to painting again, which invalidates the state, etc.

Reply:
actually, i take that back.
key is created prior to painting, so that part of equation is fine
that said, the state can still mutate during painting, invalidating the key
but it not something new that you introduces